AEPMessaging/Sources/MessagingPropositionInteraction.swift

```diff
@@ -0,0 +1,111 @@
+ OF ANY KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ */
+
+import AEPServices
+import Foundation
+
+struct MessagingPropositionInteraction: Codable {
+    var eventType: MessagingEdgeEventType
+    var interaction: String?
+    var propositionInfo: PropositionInfo
+    var itemId: String?
+
+    var xdm: [String: Any] {
+        var propositionDetailsData: [String: Any] = [:]
+
+        propositionDetailsData = [
+            MessagingConstants.XDM.Inbound.Key.ID: propositionInfo.id,
+            MessagingConstants.XDM.Inbound.Key.SCOPE: propositionInfo.scope,
+            MessagingConstants.XDM.Inbound.Key.SCOPE_DETAILS: propositionInfo.scopeDetails.asDictionary() ?? [:]
+        ]
+
+        if
+            let itemId = itemId,
+            !itemId.isEmpty
+        {
+            propositionDetailsData[MessagingConstants.XDM.Inbound.Key.ITEMS] = [
+                [
+                    MessagingConstants.XDM.Inbound.Key.ID: itemId
+                ]
+            ]
+        }
+
+        let propositionEventType: [String: Any] = [
+            eventType.propositionEventType: 1
+        ]
+
+        var decisioning: [String: Any] = [
+            MessagingConstants.XDM.Inbound.Key.PROPOSITION_EVENT_TYPE: propositionEventType,
+            MessagingConstants.XDM.Inbound.Key.PROPOSITIONS: [propositionDetailsData]
+        ]
+
+        // only add `propositionAction` data if this is an interact event
+        if
+            eventType == .interact,
+            let interaction = interaction
+        {
+            let propositionAction: [String: String] = [
+                MessagingConstants.XDM.Inbound.Key.ID: interaction,
+                MessagingConstants.XDM.Inbound.Key.LABEL: interaction
+            ]
+            decisioning[MessagingConstants.XDM.Inbound.Key.PROPOSITION_ACTION] = propositionAction
+        }
+
+        let experience: [String: Any] = [
+            MessagingConstants.XDM.Inbound.Key.DECISIONING: decisioning
+        ]
+
+        let xdm: [String: Any] = [
+            MessagingConstants.XDM.Key.EVENT_TYPE: eventType.toString(),
+            MessagingConstants.XDM.AdobeKeys.EXPERIENCE: experience
+        ]
+        return xdm
+    }
+
+    enum CodingKeys: String, CodingKey {
+        case eventType
+        case interaction
+        case propositionInfo
+        case itemId
+    }
+
+    init(eventType: MessagingEdgeEventType, interaction: String?, propositionInfo: PropositionInfo, itemId: String?) {
+        self.eventType = eventType
+        self.interaction = interaction
+        self.propositionInfo = propositionInfo
+        self.itemId = itemId
+    }
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: CodingKeys.self)
+
+        try container.encode(eventType.toString(), forKey: .eventType)
```

Comment:
are we storing/retrieving from string instead of int to account for the deprecation of older enum values?

Reply:
Yeah, the raw values are different.